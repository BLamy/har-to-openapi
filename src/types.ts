import type { OpenApiSpec } from "@loopback/openapi-v3-types";

export interface HarToOpenAPIConfig {
  /** if true, we'll treat every url as having the same domain, regardless of what its actual domain is
    * the first domain we see is the domain we'll use
    * @defaultValue `false` */
  forceAllRequestsInSameSpec?: boolean;
  /** if true, every path object will have its own servers entry, defining its base path. This is useful when
    * forceAllRequestsInSameSpec is set */
  addServersToPaths?: boolean;
  /** try and guess common auth headers
    * @defaultValue `true` */
  guessAuthenticationHeaders?: boolean;
  /** if the response has this status code, ignore the body */
  ignoreBodiesForStatusCodes?: number[];
  /** whether non standard methods should be allowed (like HTTP MY_CUSTOM_METHOD)
    * @defaultValue `false` */
  relaxedMethods?: boolean;
  /** whether we should try and parse non application/json responses as json
    * @defaultValue `true` */
  relaxedContentTypeJsonParse?: boolean;
  /** a list of tags that match passed on the path, either [match_and_tag] or [match, tag] */
  tags?: ([string] | [string, string] | string)[] | ((url: string) => string | string[] | void);
  /** response mime types to filter for */
  mimeTypes?: string[];
  /** known security headers for this har, to add to security field in openapi (e.g. "X-Auth-Token") */
  securityHeaders?: string[];
  /** Whether to filter out all standard headers from the parameter list in openapi
    * @defaultValue `true` */
  filterStandardHeaders?: boolean;
  /** Whether to log errors to console
    * @defaultValue `false` */
  logErrors?: boolean;
  /** a string, regex, or callback to filter urls for inclusion */
  urlFilter?: string | RegExp | ((url: string) => boolean | Promise<boolean>);
  /** when we encounter a URL, try and parameterize it, such that something like
    * GET /uuids/123e4567-e89b-12d3-a456-426655440000 becomes GET /uuids/{uuid}
    * @defaultValue `false` */
  attemptToParameterizeUrl?: boolean;
  /** when we encounter a path without a response or with a response that does not have 2xx, dont include it
    * @defaultValue `false` */
  dropPathsWithoutSuccessfulResponse?: boolean;
}

type Required<T, K extends keyof T> = T & { [P in K]-?: T[P] };
type WithRequired<T, K extends keyof T> = Pick<T, Exclude<keyof T, K>> & Required<T, K>;

export type InternalConfig = WithRequired<
  HarToOpenAPIConfig,
  | "filterStandardHeaders"
  | "addServersToPaths"
  | "attemptToParameterizeUrl"
  | "relaxedMethods"
  | "guessAuthenticationHeaders"
  | "dropPathsWithoutSuccessfulResponse"
  | "relaxedContentTypeJsonParse"
  | "forceAllRequestsInSameSpec"
>;

export interface HarToOpenAPISpec {
  spec: OpenApiSpec;
  yamlSpec: string;
  domain: string | undefined;
}
