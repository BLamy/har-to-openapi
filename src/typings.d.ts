type JSONSchema = object;

declare module "json-schema-deref-sync" {
  declare function index(schema: JSONSchema, options?: json_schema_deref_sync.Options): JSONSchema;
  interface Options {
    baseFolder?: string;
    failOnMissing?: boolean;
    loaders?: {
      [key: string]: (reference: string, options?: Options) => JSONSchema;
    };
    mergeAdditionalProperties?: boolean;
    removeIds?: boolean;
  }

  export = index;
}
