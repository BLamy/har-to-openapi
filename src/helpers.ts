import type { InternalConfig } from "./types";
import type {
  HeadersObject,
  OperationObject,
  ParameterObject,
  RequestBodyObject,
  ResponseObject,
  SchemaObject,
  SecurityRequirementObject,
} from "@loopback/openapi-v3-types";
import type { Content, Cookie, Header, PostData, PostDataParams, QueryString, Response } from "har-format";
import toOpenApiSchema from "@openapi-contrib/json-schema-to-openapi-schema";
import { quicktypeJSON } from "./quicktype";
import { camelCase, cloneDeep, startCase, uniqBy } from "lodash-es";
import { shouldFilterHeader } from "./utils/headers";
import { URLSearchParams } from "url";
import { getCookieSecurityName, getTypenameFromPath } from "./utils/string";

export const addMethod = (method: string, url: URL, config: InternalConfig): OperationObject => {
  const path = url.pathname;
  // generate operation id
  const summary = `${method} ${getTypenameFromPath(path)}`;
  const operationId = camelCase(summary);
  const tags = config?.tags || [];
  let pathTags: string[] = [];
  if (typeof tags === "function") {
    const userDefinedTags = tags(path);
    pathTags = [userDefinedTags || []].flat();
  } else {
    for (const tag of tags) {
      const isTagArray = Array.isArray(tag);
      const comparison = isTagArray ? tag[0] : tag;
      if (path.includes(comparison)) {
        const tagToApply: string = isTagArray ? (tag.length === 2 ? tag[1] : tag[0]) : tag;
        pathTags.push(tagToApply);
      }
    }
  }

  const operationsObject = {
    operationId,
    description: "",
    summary: startCase(summary),
    parameters: [],
    responses: {},
  } as OperationObject;
  if (config?.addServersToPaths) {
    const server = {
      url: url.origin,
    };
    operationsObject.servers = [server]; // not perfect but we can try and set the servers property here
  }
  if (pathTags?.length) {
    operationsObject.tags = pathTags;
  }
  return operationsObject;
};

export const addRequestHeaders = (specMethod: OperationObject, headers: Header[], config: InternalConfig) => {
  const parameters = (specMethod.parameters ??= []);
  const { filterStandardHeaders, securityHeaders = [] } = config;
  const customHeaders = filterStandardHeaders
    ? headers.filter((header) => {
        return !shouldFilterHeader(header.name, securityHeaders);
      })
    : headers;
  customHeaders.forEach((header) => {
    parameters.push({
      schema: {
        type: "string",
        default: header.value,
        example: header.value,
      },
      in: "header",
      name: header.name,
      description: header.name,
    } as ParameterObject);
  });
  specMethod.parameters = uniqBy(parameters, (elem: any) => {
    return `${elem.name}:${elem.in}:${elem.$ref}`;
  });
};

export const addQueryStringParams = (specMethod: OperationObject, harParams: QueryString[]) => {
  const methodQueryParameters: string[] = [];
  const parameters = (specMethod.parameters ??= []);
  parameters.forEach((param) => {
    if ("in" in param && param.in === "query") {
      methodQueryParameters.push(param.name);
    }
  });
  harParams?.forEach((param) => {
    if (!methodQueryParameters.includes(param.name)) {
      // add query parameter
      parameters.push({
        schema: {
          type: "string",
          default: decodeURIComponent(param.value),
          example: decodeURIComponent(param.value),
        },
        in: "query",
        name: param.name,
        description: param.name,
      });
    }
  });
};

export const getSecurity = (
  headers: Header[],
  securityHeaders: string[],
  cookies: Cookie[] | undefined,
): SecurityRequirementObject | undefined => {
  const security: SecurityRequirementObject = {};
  headers.forEach(function (header) {
    const headerName = header.name.trim().toLowerCase();
    if (securityHeaders.includes(headerName)) {
      security[header.name] = [];
      if (headerName === "cookie" && cookies?.length) {
        cookies.forEach((cookie) => {
          const securityName = getCookieSecurityName(cookie);

          security["cookie"]?.push(securityName);
        });
      }
    }
  });
  if (Object.keys(security).length === 0) {
    return undefined;
  }
  return security;
};

const mapParams = (params: PostDataParams["params"]): SchemaObject => {
  const properties: SchemaObject["properties"] = {};
  const required: SchemaObject["required"] = params.map((query) => {
    if (query.value == "" || query.value == "(binary)") {
      properties[query.name] = {
        type: "string",
        format: "binary",
      };
      return query.name;
    }
    properties[query.name] = {
      type: "string",
    };
    return query.name;
  });
  return {
    type: "object",
    properties,
    required,
  };
};
const getFormData = (postData: PostData | Content): SchemaObject | undefined => {
  if (postData && "params" in postData && postData?.params?.length && postData.params.length > 0) {
    return mapParams(postData.params);
  }
  if (postData && "text" in postData) {
    const searchParams = new URLSearchParams(postData.text);
    const params: PostDataParams["params"] = [];
    searchParams.forEach((value, key) => {
      params.push({ value, name: key });
    });
    return mapParams(params);
  }
};

export const getBody = async (
  postData: PostData | Content | undefined,
  details: { urlPath: string; method: string; examples: any[] },
  config: InternalConfig,
): Promise<RequestBodyObject | undefined> => {
  if (!postData || !postData.mimeType) {
    return undefined;
  }

  const { urlPath, method, examples } = details;

  const param: RequestBodyObject = {
    required: true,
    content: {},
  };
  const text = postData.text;
  const options = {
    cloneSchema: true,
    dereference: true,
    dereferenceOptions: {
      dereference: {
        circular: "ignore",
      },
    },
  } as Parameters<typeof toOpenApiSchema>[1];
  if (postData && text !== undefined) {
    const mimeTypeWithoutExtras = postData.mimeType.split(";")[0];
    const string = mimeTypeWithoutExtras?.split("/").filter(Boolean);
    const mime = string.pop();
    // We run the risk of circular references here
    const mimeLower = mime!.toLocaleLowerCase();
    switch (mimeLower) {
      case "form-data":
      case "x-www-form-urlencoded":
        const formSchema = getFormData(postData);
        if (formSchema) {
          const schema = await toOpenApiSchema(formSchema, options);
          if (schema) {
            // @ts-ignore
            param.content[mimeTypeWithoutExtras] = {
              schema,
            };
          }
        }
        break;
      // try and parse plain and text as json as well
      case "plain":
      case "text":
      case "json":
      default:
        if (mimeLower === "json" || config?.relaxedContentTypeJsonParse) {
          try {
            const isBase64Encoded = "encoding" in postData && (<any>postData).encoding == "base64";
            const data = JSON.parse(isBase64Encoded ? Buffer.from(text, "base64").toString() : text);
            examples.push(JSON.stringify(data));
            const typeName = camelCase([getTypenameFromPath(urlPath), method, "request"].join(" "));
            const jsonSchema = await quicktypeJSON("schema", typeName, examples);
            const schema = await toOpenApiSchema(cloneDeep(jsonSchema), options);
            param.content[mimeTypeWithoutExtras] = {
              // @ts-ignore
              schema,
              example: data,
            };
          } catch (err) {
            // do nothing on json parse failures
          }
        }
        break;
    }
  } else {
    const multipartMimeType = "multipart/form-data";
    // Don't apply the fallback to if there is a filter and it doesn't include it
    if (!config.mimeTypes || config.mimeTypes.includes(multipartMimeType)) {
      param.content = {
        [multipartMimeType]: {
          schema: {
            properties: {
              filename: {
                description: "",
                format: "binary",
                type: "string",
              },
            },
            type: "object",
          },
        },
      };
    }
  }
  return param;
};

export const getResponseBody = async (
  response: Response,
  details: { urlPath: string; method: string; examples: any[] },
  config: InternalConfig,
): Promise<ResponseObject | undefined> => {
  const body = await getBody(response.content, details, config);

  const { filterStandardHeaders, securityHeaders } = config;
  const param: ResponseObject = {
    description: "",
  };
  if (body?.content) {
    param.content = body.content;
  }

  const headers = response.headers || [];
  const customHeaders = filterStandardHeaders
    ? headers.filter((header) => {
        return !shouldFilterHeader(header.name, securityHeaders);
      })
    : headers;
  if (customHeaders.length) {
    param.headers = customHeaders.reduce<HeadersObject>((acc, header) => {
      acc[header.name] = {
        description: `Custom header ${header.name}`,
        schema: {
          type: "string",
        },
      };
      return acc;
    }, {} as HeadersObject);
  }
  if (param.headers || param.content) {
    return param;
  }
};
