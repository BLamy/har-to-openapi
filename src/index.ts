import type { OperationObject } from "@loopback/openapi-v3-types";
import { createEmptyApiSpec } from "@loopback/openapi-v3-types";
import type { Entry, Har, QueryString } from "har-format";
import YAML from "js-yaml";
import sortJson from "sort-json";
import { addMethod, addQueryStringParams, addRequestHeaders, getBody, getResponseBody, getSecurity } from "./helpers";
import type { Config, IGenerateSpecResponse, InternalConfig } from "./types";
import type {
  PathItemObject,
  PathsObject,
  SecurityRequirementObject,
  ServerObject,
} from "openapi3-ts/src/model/OpenApi";
import { cloneDeep, groupBy } from "lodash-es";
import { addResponse } from "./utils/baseResponse";
import { isStandardMethod } from "./utils/methods";

const checkPathFromFilter = async (urlPath: string, filter: Config["urlFilter"]) => {
  if (typeof filter === "string") {
    return urlPath.includes(filter);
  }
  if (filter instanceof RegExp) {
    return filter.test(urlPath);
  }
  if (typeof filter === "function") {
    return filter(urlPath);
  }
};

const getConfig = (config?: Config): InternalConfig => {
  const internalConfig = cloneDeep(config || {}) as InternalConfig;
  // set up some defaults
  internalConfig.filterStandardHeaders ??= true;
  internalConfig.relaxedContentTypeJsonParse ??= true;
  internalConfig.forceAllRequestsInSameSpec ??= false;
  internalConfig.relaxedMethods ??= false;
  internalConfig.logErrors ??= false;
  if (internalConfig.securityHeaders) {
    internalConfig.securityHeaders = internalConfig.securityHeaders.map((l) => l.toLowerCase());
  }
  return internalConfig;
};

function tryGetHostname(url: string, logErrors: boolean | undefined, fallback: string): string;
function tryGetHostname(url: string, logErrors: boolean | undefined): string | undefined;
function tryGetHostname(url: string, logErrors?: boolean, fallback?: string): string | undefined {
  try {
    return new URL(url).hostname;
  } catch {
    if (logErrors) {
      console.error(`Error parsing url ${url}`);
    }
  }
  return fallback;
}

const generateSpecs = async <T extends Har>(har: T, config?: Config): Promise<IGenerateSpecResponse[]> => {
  if (!har?.log?.entries?.length) {
    return [];
  }

  // decode base64 now
  har.log.entries.forEach((item) => {
    const response = item.response;
    if (response && response.content?.encoding === "base64") {
      response.content.text = Buffer.from(response.content.text || "", "base64").toString();
      delete response.content.encoding;
    }
  });

  const internalConfig = getConfig(config);
  const {
    ignoreBodiesForStatusCodes,
    mimeTypes,
    securityHeaders,
    filterStandardHeaders,
    forceAllRequestsInSameSpec,
    urlFilter,
    relaxedMethods,
    logErrors,
  } = internalConfig;

  const groupedByHostname = groupBy(har.log.entries, (entry: Entry) => {
    if (forceAllRequestsInSameSpec) {
      return "specs";
    }
    return tryGetHostname(entry.request.url, logErrors);
  });
  const specs: IGenerateSpecResponse[] = [];

  for (const domain in groupedByHostname) {
    try {
      // loop through har entries
      const spec = createEmptyApiSpec();
      spec.info.title = "HarToOpenApi";

      const harEntriesForDomain = groupedByHostname[domain];

      const securitySchemas: SecurityRequirementObject[] = [];
      const firstUrl = harEntriesForDomain[0].request.url;

      for (const item of harEntriesForDomain) {
        try {
          const url = item.request.url;

          if (!url) {
            continue;
          }

          // filter and collapse path urls
          const urlObj = new URL(url);
          const urlPath = urlObj.pathname;

          const queryParams = urlObj.search;

          if (urlFilter) {
            const isValid = await checkPathFromFilter(urlObj.href, urlFilter);
            if (!isValid) {
              continue;
            }
          }
          const mimeType = item.response?.content?.mimeType;
          const isValidMimetype = !mimeTypes || (mimeType && mimeTypes.includes(mimeType));
          if (!isValidMimetype) {
            continue;
          }

          // create method
          const method = item.request.method.toLowerCase();
          // if its not standard and we're not in relaxed mode, skip it
          if (!relaxedMethods && !isStandardMethod(method)) {
            continue;
          }
          // create path if it doesn't exist
          spec.paths[urlPath] ??= { parameters: [] } as PathsObject;
          const path = spec.paths[urlPath] as PathItemObject;

          path[method] ??= addMethod(method, urlObj, config);
          const specMethod = path[method] as OperationObject;
          // generate response
          const status = item.response?.status;
          if (status) {
            specMethod.responses[status] ??= addResponse(status, method);
          }

          const requestHeaders = item.request.headers;
          if (securityHeaders?.length && requestHeaders?.length) {
            const security = getSecurity(requestHeaders, securityHeaders);
            if (security) {
              securitySchemas.push(security);
              specMethod.security = [security];
            }
          }

          // add query string parameters
          if (item.request.queryString?.length) {
            addQueryStringParams(specMethod, item.request.queryString);
          }
          if (queryParams) {
            // try and parse from the url if the har is malformed
            const queryStrings: QueryString[] = [];
            for (const entry of urlObj.searchParams.entries()) {
              queryStrings.push({ name: entry[0], value: entry[1] });
            }
            addQueryStringParams(specMethod, queryStrings);
          }
          if (requestHeaders?.length) {
            addRequestHeaders(specMethod, requestHeaders, filterStandardHeaders);
          }

          const shouldUseRequestAndResponse =
            !ignoreBodiesForStatusCodes || !ignoreBodiesForStatusCodes.includes(status);
          if (shouldUseRequestAndResponse && item.request.postData) {
            specMethod.examples ??= [];
            specMethod.requestBody = await getBody(
              item.request.postData,
              { urlPath, method, examples: specMethod.examples },
              internalConfig,
            );
          }

          if (status && isValidMimetype && shouldUseRequestAndResponse && item.response) {
            specMethod.responseExamples ??= {};
            specMethod.responseExamples[status] ??= [];
            const body = await getResponseBody(
              item.response,
              { urlPath, method, examples: specMethod.responseExamples[status] },
              internalConfig,
            );
            if (body) {
              specMethod.responses[status] = body;
            }
          }
        } catch (e) {
          if (logErrors) {
            console.error(`Error parsing ${item.request}`);
            console.error(e);
          }
          // error parsing one entry, move on
        }
      }
      // If there were no valid paths, bail
      if (!Object.keys(spec.paths).length) {
        continue;
      }

      if (securitySchemas.length) {
        spec.components ||= {};
        spec.components.securitySchemes = {};
        securitySchemas.forEach((schema) => {
          const schemaName = Object.keys(schema)[0];
          spec.components!.securitySchemes![schemaName] = {
            type: "http",
            name: schemaName,
            in: "header",
          };
        });
      }

      // remove the examples that we used to build the superset schemas
      for (const path of Object.values(spec.paths)) {
        for (const method of Object.values<OperationObject>(path)) {
          delete method.responseExamples;
          delete method.examples;
        }
      }
      // sort paths
      spec.paths = sortJson(spec.paths, { depth: 200 });
      const yamlSpec = YAML.dump(spec);
      const labeledDomain = tryGetHostname(firstUrl, logErrors, domain);
      const prefix = firstUrl?.startsWith("https://") ? "https://" : "http://";
      const server: ServerObject = {
        url: `${prefix}${labeledDomain}`,
      };
      spec.servers = [server];
      specs.push({ spec, yamlSpec, domain: labeledDomain });
    } catch (err) {
      if (logErrors) {
        console.error(`Error creating spec for ${domain}`);
      }
    }
  }

  return specs;
};
const generateSpec = async <T extends Har>(har: T, config?: Config): Promise<IGenerateSpecResponse> => {
  const specs = await generateSpecs(har, config);
  if (specs.length) {
    return specs[0];
  }
  const spec = createEmptyApiSpec();
  spec.info.title = "HarToOpenApi - no valid specs found";
  return { spec, yamlSpec: YAML.dump(spec), domain: undefined };
};

export { generateSpec, generateSpecs };
