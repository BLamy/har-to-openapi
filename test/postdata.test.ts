import { describe, test } from "vitest";
import OpenAPISchemaValidator from "openapi-schema-validator";
import { generateSpec } from "../src";
import type { Har } from "har-format";

const validator = new OpenAPISchemaValidator({
  version: 3,
});
const har = {
  log: {
    entries: [
      {
        index: 0,
        request: {
          method: "POST",
          url: "http://test.loadimpact.com/login",
          headers: [
            {
              name: "Content-Type",
              value: "application/json",
            },
          ],
          postData: {
            mimeType: "application/json",
            text: '{"user":"admin","password":"123"}',
          },
        },
      },
    ],
  },
} as unknown as Har;
describe("har-to-openapi", () => {
  test(`HAR with request postdata matches snapshot`, async ({ expect }) => {
    expect(await generateSpec(har)).toMatchSnapshot();
  });
  test(`HAR with request postdata has correct options`, async ({ expect }) => {
    const { spec } = await generateSpec(har);
    const result = validator.validate(spec as any);
    expect(result.errors).toHaveLength(0);

    const loginPath = spec.paths["/login"];
    expect(loginPath).toBeDefined();
    expect(loginPath).toHaveProperty("post");
    const post = loginPath["post"];
    expect(post).toHaveProperty("requestBody");
    const requestBody = post.requestBody;
    expect(requestBody).toHaveProperty("content");
    expect(requestBody.content).toHaveProperty("application/json");
    const contentElement = requestBody.content["application/json"];
    expect(contentElement).toHaveProperty("schema");
    const schema = contentElement.schema;
    expect(schema).toHaveProperty("type");
    expect(schema.type).toBe("object");
    expect(schema).toHaveProperty("properties");
    expect(schema.properties).toHaveProperty("user");
    expect(schema.properties).toHaveProperty("password");
    expect(schema.properties.user).toHaveProperty("type");
    expect(schema.properties.user.type).toBe("string");
    expect(contentElement).toHaveProperty("example");
    expect(contentElement.example.user).toBe("admin");
    expect(schema.properties.password).toHaveProperty("type");
    expect(schema.properties.password.type).toBe("string");
    expect(contentElement.example.password).toBe("123");
  });
});
