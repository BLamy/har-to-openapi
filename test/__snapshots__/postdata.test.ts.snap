// Vitest Snapshot v1

exports[`har-to-openapi > HAR with request postdata matches snapshot 1`] = `
{
  "spec": {
    "info": {
      "title": "HarToOpenApi",
      "version": "1.0.0",
    },
    "openapi": "3.0.0",
    "paths": {
      "/login": {
        "parameters": [],
        "post": {
          "description": "",
          "operationId": "post-login",
          "parameters": [],
          "requestBody": {
            "content": {
              "application/json": {
                "example": {
                  "password": "123",
                  "user": "admin",
                },
                "schema": {
                  "additionalProperties": false,
                  "properties": {
                    "password": {
                      "format": "integer",
                      "type": "string",
                    },
                    "user": {
                      "type": "string",
                    },
                  },
                  "required": [],
                  "title": "LoginPostRequest",
                  "type": "object",
                },
              },
            },
            "required": true,
          },
          "responses": {},
          "summary": "Create ",
          "tags": [
            "",
          ],
        },
      },
    },
    "servers": [
      {
        "url": "/",
      },
    ],
  },
  "yamlSpec": "openapi: 3.0.0
info:
  title: HarToOpenApi
  version: 1.0.0
paths:
  /login:
    parameters: []
    post:
      description: ''
      operationId: post-login
      parameters: []
      requestBody:
        content:
          application/json:
            example:
              password: '123'
              user: admin
            schema:
              additionalProperties: false
              properties:
                password:
                  format: integer
                  type: string
                user:
                  type: string
              required: []
              title: LoginPostRequest
              type: object
        required: true
      responses: {}
      summary: 'Create '
      tags:
        - ''
servers:
  - url: /
",
}
`;
